### ISSUE TITLE
Master fails tests with JDK-17

### ISSUE THREAD

**Issue by mouse07410** [bcgit/bc-java#1107]
macOS Monterey 12.2.1 (probably irrelevant), JDK-17 (Amazon Corretto, probably irrelevant), current BC master:

![](https://user-images.githubusercontent.com/5923577/153594421-f7b9ddc0-aaf2-4cdd-9ef3-b3c7493b2900.png)

I observe it consistently on a machine that only has JDK-17 (Amazon Corretto, if it matters - but I doubt that). Machine that builds and tests with corretto-11 and has corretto-17 installed, does not produce any errors.
```
CertTest: 9 failed - exception java.security.InvalidKeyException: cannot identify EdDSA public key 
at org.bouncycastle.cert.test.AllTests.testSimpleTests(AllTests.java:30) 
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) 
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) 
at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) 
at junit.extensions.TestDecorator.basicRun(TestDecorator.java:23) 
at junit.extensions.TestSetup$1.protect(TestSetup.java:23) 
at junit.extensions.TestSetup.run(TestSetup.java:27) 
.  .  . 
index 9 BlockCipher: IDEA/PGPCFB/NoPadding failed encryption - java.io.IOException: javax.crypto.ShortBufferException: output buffer too short for input. 

junit.framework.AssertionFailedError: index 9 BlockCipher: IDEA/PGPCFB/NoPadding failed encryption - java.io.IOException: javax.crypto.ShortBufferException: output buffer too short for input. 
at org.bouncycastle.jce.provider.test.SimpleTestTest.testJCE(SimpleTestTest.java:26) 
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) 
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) 
at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) 
at junit.extensions.TestDecorator.basicRun(TestDecorator.java:23) 
at junit.extensions.TestSetup$1.protect(TestSetup.java:23) 
at junit.extensions.TestSetup.run(TestSetup.java:27)
.  .  .
Expected: <null> but was: junit.framework.AssertionFailedError 

junit.framework.AssertionFailedError: Expected: <null> but was: junit.framework.AssertionFailedError 
at org.bouncycastle.jsse.provider.test.TestProtocolUtil.runClientAndServer(TestProtocolUtil.java:68) 
at org.bouncycastle.jsse.provider.test.EdDSACredentialsTest.implTestEdDSACredentials(EdDSACredentialsTest.java:211)
at org.bouncycastle.jsse.provider.test.EdDSACredentialsTest.test12_Ed25519(EdDSACredentialsTest.java:181) 
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) 
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) 
at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) 
at junit.extensions.TestDecorator.basicRun(TestDecorator.java:23) 
at junit.extensions.TestSetup$1.protect(TestSetup.java:23) 
at junit.extensions.TestSetup.run(TestSetup.java:27) 
.  .  .
Expected: <null> but was: junit.framework.AssertionFailedError 

junit.framework.AssertionFailedError: Expected: <null> but was: junit.framework.AssertionFailedError 
at org.bouncycastle.jsse.provider.test.TestProtocolUtil.runClientAndServer(TestProtocolUtil.java:68) 
at org.bouncycastle.jsse.provider.test.EdDSACredentialsTest.implTestEdDSACredentials(EdDSACredentialsTest.java:211)
at org.bouncycastle.jsse.provider.test.EdDSACredentialsTest.test12_Ed448(EdDSACredentialsTest.java:186) 
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) 
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) 
at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) 
at junit.extensions.TestDecorator.basicRun(TestDecorator.java:23) 
at junit.extensions.TestSetup$1.protect(TestSetup.java:23) 
at junit.extensions.TestSetup.run(TestSetup.java:27) 
.  .  .
Expected: <null> but was: junit.framework.AssertionFailedError 

junit.framework.AssertionFailedError: Expected: <null> but was: junit.framework.AssertionFailedError 
at org.bouncycastle.jsse.provider.test.TestProtocolUtil.runClientAndServer(TestProtocolUtil.java:68) 
at org.bouncycastle.jsse.provider.test.EdDSACredentialsTest.implTestEdDSACredentials(EdDSACredentialsTest.java:211)
at org.bouncycastle.jsse.provider.test.EdDSACredentialsTest.test13_Ed25519(EdDSACredentialsTest.java:191) 
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) 
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) 
at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) 
at junit.extensions.TestDecorator.basicRun(TestDecorator.java:23) 
at junit.extensions.TestSetup$1.protect(TestSetup.java:23) 
at junit.extensions.TestSetup.run(TestSetup.java:27)
.  .  .
Expected: <null> but was: junit.framework.AssertionFailedError 

junit.framework.AssertionFailedError: Expected: <null> but was: junit.framework.AssertionFailedError 
at org.bouncycastle.jsse.provider.test.TestProtocolUtil.runClientAndServer(TestProtocolUtil.java:68) 
at org.bouncycastle.jsse.provider.test.EdDSACredentialsTest.implTestEdDSACredentials(EdDSACredentialsTest.java:211)
at org.bouncycastle.jsse.provider.test.EdDSACredentialsTest.test13_Ed448(EdDSACredentialsTest.java:196) 
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) 
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) 
at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) 
at junit.extensions.TestDecorator.basicRun(TestDecorator.java:23) 
at junit.extensions.TestSetup$1.protect(TestSetup.java:23) 
at junit.extensions.TestSetup.run(TestSetup.java:27)
```
 
And it would still help to see the content of `bnd.sh`. 
Right now I don't run `jar_process.sh` because it relies upon JDK-9 that I don't have (and don't intend to), and uses JDK-15 (which I don't have) instead of LTS JDK-17.

**Comment by dghgit** [comment]
You can replace everything in jar_process.sh with JDK 17 - what's used in jar_process.sh reflects the minimum required for a particular stage to work, just comment out the call to bnd.sh for now, rewriting the manifest for OSGI will probably just make things more confusing. It will be easy to add in later.

**Comment by mouse07410** [comment]
> You can replace everything in jar_process.sh with JDK 17 - what's used in jar_process.sh reflects the minimum required for a particular stage to work, 

`jar_process.sh`, thankfully, is simple.

My _real_ problem is with `add_module.sh` file. It 
seems to be taking stuff from what's built by JDK-9, like
```bash
    export JAVA_HOME=$JAVA_11
    export PATH=$JAVA_HOME/bin:$PATH 

    cd module.tmp
    jar uf $jarName --release 9 -C v9 .
    if [ -d v11 ]
    then
        jar uf $jarName --release 11 -C v11 .
    fi
```

I want to convert it to JDK11 and JDK17 on one machine, and JDK17-only  on the other.

> just comment out the call to bnd.sh for now, rewriting the manifest for OSGI will probably just make things more confusing. It will be easy to add in later.

Yes, I agree. Will do. OSGI is not the primacy concern right now.

>  if you under src/main and src/test along with java you will also see jdk1.9, jdk1.11, jdk1.15 - these are there because of the multi-release. If you overlay jdk1.9 onto java, then overlay jdk1.11 onto java, and then overlay jdk1.15 onto java you'll end up with a source tree for jdk1.15, which might also give you what you want.

I'm sorry, I don't think I understand. Here's what I have for `prov/src/main`:
```
$ exa prov/src/main
ext-jdk1.9  java  javadoc  jdk1.1  jdk1.2  jdk1.3  jdk1.4  jdk1.9  jdk1.11  jdk1.15  resources
$ 
```
Overlay - how? Copy files from `jdk1.15/` to `java/`? 
I'd rather not muck with the contents of the existing directories, and stay compatible with what's on GitHub, so I don't need to do a ton of manual updates when stuff in those directories (including `java/`) changes.

**Comment by dghgit** [comment]
Yes, copy. One option might be to have your on gradle file for java11/java17 and java17 only and then just blow away everything else when you update.

With add_module.sh - as I mentioned earlier the specified JDKs are there as  _**the minimum required**_ so you can replace them with later ones quite safely. I'd suggest ditching  the use of --release for Java 11 and below since that's your minimum. In this case you'd just overlay the classes directly (instead of the source code) rather than dump them in the versions tree (note this means making sure they go in with org at the start, not v9 for example). I'd recommend this as I'm not sure what the outcome would be like if you put java 11 byte-code into a java-9 versions directory, it might work in your case now, but you may also find yourself burned by it later.

**Comment by mouse07410** [comment]
> Yes, copy. One option might be to have your on gradle file for java11/java17 and java17 only and then just blow away everything else when you update

Frankly, this is the least attractive option. I prefer to avoid it.

> With `add_module.sh` - as I mentioned earlier the specified JDKs are there as **_the minimum required_** so you can replace them with later ones quite safely. 

I'd love to see an example. Because my hacked script doesn't seem to compile and/or add `module-info` to the provider jars.

> I'd suggest ditching the use of --release for Java 11 and below since that's your minimum.

On _one_ machine Java-11 is the minimum. On another one - Java-17 is the minimum. I'd like the same script to handle both.

> I'd recommend this as I'm not sure what the outcome would be like if you put java 11 byte-code into a java-9 versions directory, it might work in your case now, but you may also find yourself burned by it later.

Ideally, the script should check for presence of Java-9, Java-11, Java-17 - and act accordingly. Guidance on how best to do it is welcome. 

Here's the (insufficiently) hacked `add_module.sh` for the machine with only JDK-17 (BTW, I copied the entire content of `jdk1.15/` to `jdk1.17/`):
```bash
export JAVA_9=""
#export JAVA_11=`/usr/libexec/java_home -v 11`
export JAVA_11=""
export JAVA_17=`/usr/libexec/java_home -v 17`

if [ ! -z $JAVA_9 ]; then
  export JAVA_HOME=$JAVA_9
  MINTARGET="1.9"
elif [ ! -z $JAVA_11 ]; then
  export JAVA_HOME=$JAVA_11
  MINTARGET="11"
elif [ ! -z $JAVA_17 ]; then
  export JAVA_HOME=$JAVA_17
  MINTARGET="17"
else
  echo "No acceptable JDK found - exiting..."
  exit 0
fi
export PATH=$JAVA_HOME/bin:$PATH

echo "JAVA_9=${JAVA_9}"
echo "JAVA_11=${JAVA_11}"
echo "JAVA_17=${JAVA_17}"
echo "PATH=${PATH}"

if [ $# -ne 1 ]
then
   echo "usage: add_module.sh prov|prov-ext|util|tls|pg|pkix|mail|jmail"
   exit 0
fi

for jarName in bc$1-jdk18on-*.jar
do

rm -rf module.tmp

# Java 9 Step
(

    if [ ! -z $JAVA_9 ]; then
      export JAVA_HOME=$JAVA_9
    fi
    export PATH=$JAVA_HOME/bin:$PATH

    mkdir -p module.tmp/v5
    mkdir -p module.tmp/versions/v9
    mkdir -p module.tmp/versions/v11
    mkdir -p module.tmp/versions/v17
    ( cd module.tmp/v5; jar xf ../../build/artifacts/jdk1.8/jars/$jarName )

    provJar=`echo build/artifacts/jdk1.8/jars/bcprov-jdk18on*.jar`
    utilJar=`echo build/artifacts/jdk1.8/jars/bcutil-jdk18on*.jar`
    pkixJar=`echo build/artifacts/jdk1.8/jars/bcpkix-jdk18on*.jar`
    if [ $1 = "prov" ]
    then
         if [ ! -z $JAVA_11 ]; then
           echo "Building prov for $JAVA_11"
	          $JAVA_11/bin/javac -classpath module.tmp/v5 -d module.tmp/v11 `find $1/src/main/jdk1.11 -name "*.java"` -sourcepath $1/src/main/jdk1.11:$1/src/main/java:core/src/main/java
         fi
         if [ ! -z $JAVA_17 ]; then
           echo "Building prov for $JAVA_17"
	          $JAVA_17/bin/javac -classpath module.tmp/v5:module.tmp/v11 -d module.tmp/v17 `find prov/src/main/jdk1.17 -name "*.java"` -sourcepath prov/src/main/jdk1.17:prov/src/main/jdk1.11:prov/src/main/java:core/src/main/java
         fi
    elif [ $1 = "prov-ext" ]
    then
         if [ ! -z $JAVA_11 ]; then
           echo "Building prov-ext for $JAVA_11"
	          $JAVA_11/bin/javac -classpath module.tmp/v5 -d module.tmp/v11 `find prov/src/main/jdk1.11 -name "*.java"` -sourcepath prov/src/main/jdk1.11:prov/src/main/java:core/src/main/java
          fi
          if [ ! -z $JAVA_17 ]; then
            echo "Building prov-ext for $JAVA_17"
	          $JAVA_17/bin/javac -classpath module.tmp/v5:module.tmp/v11 -d module.tmp/v17 `find prov/src/main/jdk1.17 -name "*.java"` -sourcepath prov/src/main/jdk1.17:prov/src/main/jdk1.11:prov/src/main/java:core/src/main/java
          fi
    elif [ $1 = "mail" ]
    then
	       javac  -target ${MINTARGET} --module-path ${provJar}:$utilJar:$pkixJar -classpath module.tmp/v5 -d module.tmp/v9 `find $1/src/main/jdk1.9 -name "*.java"` -sourcepath $1/src/main/jdk1.9:$1/src/main/java
    elif [ $1 = "jmail" ]
    then
         jmailSrc=`echo build/artifacts/jdk1.8/bcjmail-jdk18on-*/src/`
	       javac  -target ${MINTARGET} --module-path ${provJar}:$utilJar:$pkixJar:libs/jakarta.mail-2.0.1.jar:libs/jakarta.activation-api-2.0.0.jar -classpath module.tmp/v5 -d module.tmp/v9 `find $1/src/main/jdk1.9 -name "*.java"` -sourcepath $1/src/main/jdk1.9:$jmailSrc
    elif [ $1 = "pkix" ]
    then
	       javac  -target ${MINTARGET} --module-path ${provJar}:$utilJar -classpath module.tmp/v5 -d module.tmp/v9 `find $1/src/main/jdk1.9 -name "*.java"` -sourcepath $1/src/main/jdk1.9:$1/src/main/java
    elif [ $1 = "tls" ]
    then
	     javac  -target ${MINTARGET} --module-path ${provJar}:$utilJar -classpath module.tmp/v5 -d module.tmp/v9 `find $1/src/main/jdk1.9 -name "*.java"` -sourcepath $1/src/main/jdk1.9:$1/src/main/java
    else
	     javac  -target ${MINTARGET} --module-path $provJar -classpath module.tmp/v5 -d module.tmp/v9 `find $1/src/main/jdk1.9 -name "*.java"` -sourcepath $1/src/main/jdk1.9:$1/src/main/java
    fi
)
# Java 11 Step
(
    if [ ! -z ${JAVA_11} ]; then
      export JAVA_HOME=$JAVA_11
    fi
    export PATH=$JAVA_HOME/bin:$PATH

    cd module.tmp
    if [ -d v11 ]
    then
        extras=`2>&1 jar cf $jarName -C v5 . --release 9 -C v9 . --release 11 -C v11 . | egrep Warning | sed -e "s/Warning: entry //" | sed -e "s/ contains.*$//"`
    else
        extras=`2>&1 jar cf $jarName -C v5 . --release 9 -C v9 . | egrep Warning | sed -e "s/Warning: entry //" | sed -e "s/ contains.*$//"`
    fi
    for f in `echo $extras`
    do
	       p=`echo $f | sed -e "s:META-INF/versions/9::"`
	       rm v9/$p
    done
)
#sh ./bnd.sh build/artifacts/jdk1.8/jars/$jarName
cp build/artifacts/jdk1.8/jars/$jarName module.tmp/$jarName
# Java 11 Step
(
    # Unnecessary repetition
    #export JAVA_HOME=$JAVA_11
    export PATH=$JAVA_HOME/bin:$PATH

    cd module.tmp
    if [ -d v11 ]
    then
        jar uf $jarName --release ${MINTARGET} -C v11 .
    fi
)
# Java 17 Step
(
    export JAVA_HOME=$JAVA_17
    export PATH=$JAVA_HOME/bin:$PATH

    cd module.tmp
    if [ -d v17 ]
    then
        jar uf $jarName --release 17 -C v17 .
    fi
)
cp module.tmp/$jarName build/artifacts/jdk1.8/jars/$jarName

#bcsign11 build/artifacts/jdk1.8/jars/$jarName
done
```

**Comment by dghgit** [comment]
The script is still building a multi-release jar for Java 11 for the java 11 code.

For example:

    if [ -d v11 ]
    then
        jar uf $jarName --release ${MINTARGET} -C v11 .
    fi

I don't understand why you would do this? It should probably be:

    if [ -d v11 ]
    then
        (cd v11;  jar uf $jarName .)
    fi

The script also does not appear to be trying to compile module-info.java for the provider - I cannot see any reference to jdk1.9 there. 

All you are trying do is compile some java files and put them in a jar file. This is not complicated. I'd suggest ignoring add_module.sh as it seems to be calling confusion and building something from scratch, for java 17, simply using cp and javac is enough.

**Comment by mouse07410** [comment]
> The script is still building a multi-release jar for Java 11 for the java 11 code.

My goal is to build multi-release on machine(s) where both JDK-11 and JDK-17 are present. Unless I misunderstand the purpose of multi-release, and it's unnecessary...?

> The script also does not appear to be trying to compile `module-info.java` for the provider - I cannot see any reference to jdk1.9 there.

Yes... Where is `module-info.java` compiled in the original `add_module.sh`? It is possible that I mistakenly deleted that part, but I tried to be careful not to... It's probably somewhere in the "# Java-9 step" part? A pointer or a script sample would be  appreciated, especially as it would relate to JDK-11 or 17.

> I'd suggest ignoring `add_module.sh` as it seems to be calling confusion and building something from scratch, for java 17, simply using cp and javac is enough

`add_module.sh` certainly does cause confusion on my end. However, I'm afraid I need to go multi-release, because some of my machines use both Java-11 and Java-17. Am I mistaken trying to use multi-release for this purpose?

Basically, so far I've been ignoring `jar_process.sh` - just building the jars and copying them to `/Library/Java/Extensions/`, where on MacOS all that stuff lives (though with Java-9+ I need to explicitly point Java at it via `-p /Library/Java/EXtensions`).
But since it looks like I need module info, I'm trying to get `jar_process.sh` running in my environment...

**Comment by dghgit** [comment]
module-info would have been done using the Java 1.9 code.

I think what you're after is to do a java 11 javac followed by a direct jar update of the classes in jdk1.9 and jdk1.11, then compile what's in jdk1.15 using Java 17 and add using the jar --release 17. That will give you a jar which will have its primary release as Java 11 with support for the Java 15 classes that appear in Java 17, so it be target for both your platforms.

**Comment by mouse07410** [comment]
OK. The tests keep failing on JDK-17. Here's the result:
<img width="841" alt="Screen Shot 2022-02-16 at 19 53 29" src="https://user-images.githubusercontent.com/5923577/154383169-577dbd8c-cfdc-4c66-928f-19048db804aa.png">

Here's the build-and-test command:
```
./build1-8+ clean && time bash ./build1-8+ && bash ./jar_process.sh && bc-sign && time bash ./build1-8+ test
```

Here's the modified `add_module.sh` file:
```bash
#!/usr/bin/env bash -ex

export JAVA_9="`/usr/libexec/java_home -v 9`"
export JAVA_11="`/usr/libexec/java_home -v 11`"
export JAVA_15="`/usr/libexec/java_home -v 17`"

export JAVA_HOME=$JAVA_15
export PATH=$JAVA_HOME/bin:$PATH

if [ $# -ne 1 ]
then
   echo "usage: add_module.sh prov|prov-ext|util|tls|pg|pkix|mail|jmail"
   exit 0
fi

for jarName in bc$1-jdk18on-*.jar
do

rm -rf module.tmp

# Java 9 Step
(
    #export JAVA_HOME=$JAVA_9
    #export PATH=$JAVA_HOME/bin:$PATH

    mkdir -p module.tmp/v5
    mkdir -p module.tmp/v9
    mkdir -p module.tmp/v11
    mkdir -p module.tmp/v15
    mkdir -p module.tmp/versions/v9
    mkdir -p module.tmp/versions/v11
    mkdir -p module.tmp/versions/v15
    ( cd module.tmp/v5; jar xf ../../build/artifacts/jdk1.8/jars/$jarName )

    provJar=`echo build/artifacts/jdk1.8/jars/bcprov-jdk18on*.jar`
    utilJar=`echo build/artifacts/jdk1.8/jars/bcutil-jdk18on*.jar`
    pkixJar=`echo build/artifacts/jdk1.8/jars/bcpkix-jdk18on*.jar`
    JDK9MOD=`find prov/src/main/jdk1.9 -name "*.java"`
    JDK9EXTMOD=`find prov/src/main/ext-jdk1.9 -name "*.java"`
    echo "$JDK9MOD"
    echo "$JDK9EXTMOD"
    JDK11MOD=`find prov/src/main/jdk1.11 -name "*.java"`
    JDK15MOD=`find prov/src/main/jdk1.15 -name "*.java"`
    echo "$JDK11MOD"
    echo "$JDK15MOD"
    if [ $1 = "prov" ]
    then
       javac -classpath module.tmp/v5 -d module.tmp/v9 $JDK9MOD -sourcepath $1/src/main/jdk1.9:$1/src/main/java:core/src/main/java
	     $JAVA_11/bin/javac -classpath module.tmp/v5 -d module.tmp/v11 $JDK11MOD -sourcepath $1/src/main/jdk1.11:$1/src/main/java:core/src/main/java
	     $JAVA_15/bin/javac -classpath module.tmp/v5:module.tmp/v11 -d module.tmp/v15 $JDK15MOD -sourcepath prov/src/main/jdk1.15:prov/src/main/jdk1.11:prov/src/main/java:core/src/main/java
    elif [ $1 = "prov-ext" ]
    then
	    javac -target 9 -classpath module.tmp/v5 -d module.tmp/v9 $JDK9EXTMOD -sourcepath prov/src/main/ext-jdk1.9:prov/src/main/java:core/src/main/java
	    $JAVA_11/bin/javac -classpath module.tmp/v5 -d module.tmp/v11 $JDK11MOD -sourcepath prov/src/main/jdk1.11:prov/src/main/ext-jdk1.9:prov/src/main/java:core/src/main/java
	    $JAVA_15/bin/javac -classpath module.tmp/v5:module.tmp/v11 -d module.tmp/v15 $JDK15MOD -sourcepath prov/src/main/jdk1.15:prov/src/main/jdk1.11:prov/src/main/ext-jdk1.9:prov/src/main/java:core/src/main/java
    elif [ $1 = "mail" ]
    then
	javac  -target 9 --module-path ${provJar}:$utilJar:$pkixJar -classpath module.tmp/v5 -d module.tmp/v9 $JDK9MOD -sourcepath $1/src/main/jdk1.9:$1/src/main/java
    elif [ $1 = "jmail" ]
    then
        jmailSrc=`echo build/artifacts/jdk1.8/bcjmail-jdk18on-*/src/`
	javac  -target 9 --module-path ${provJar}:$utilJar:$pkixJar:libs/jakarta.mail-2.0.1.jar:libs/jakarta.activation-api-2.0.0.jar -classpath module.tmp/v5 -d module.tmp/v9 $JDK9MOD -sourcepath $1/src/main/jdk1.9:$jmailSrc
    elif [ $1 = "pkix" ]
    then
	javac  -target 9 --module-path ${provJar}:$utilJar -classpath module.tmp/v5 -d module.tmp/v9 $JDK9MOD -sourcepath $1/src/main/jdk1.9:$1/src/main/java
    elif [ $1 = "tls" ]
    then
	javac  -target 9 --module-path ${provJar}:$utilJar -classpath module.tmp/v5 -d module.tmp/v9 $JDK9MOD -sourcepath $1/src/main/jdk1.9:$1/src/main/java
    else
	javac  -target 9 --module-path $provJar -classpath module.tmp/v5 -d module.tmp/v9 $JDK9MOD -sourcepath $1/src/main/jdk1.9:$1/src/main/java
    fi
)
# Java 11 Step
(
    export JAVA_HOME=$JAVA_11
    export PATH=$JAVA_HOME/bin:$PATH

    cd module.tmp
    if [ -d v11 ]
    then
        extras=`2>&1 jar cf $jarName -C v5 . --release 9 -C v9 . --release 11 -C v11 . | egrep Warning | sed -e "s/Warning: entry //" | sed -e "s/ contains.*$//"`
    else
        extras=`2>&1 jar cf $jarName -C v5 . --release 9 -C v9 . | egrep Warning | sed -e "s/Warning: entry //" | sed -e "s/ contains.*$//"`
    fi
    for f in `echo $extras`
    do
	p=`echo $f | sed -e "s:META-INF/versions/9::"`
	rm v9/$p
    done
)
#sh ./bnd.sh build/artifacts/jdk1.8/jars/$jarName
cp build/artifacts/jdk1.8/jars/$jarName module.tmp/$jarName
# Java 11 Step
(
    export JAVA_HOME=$JAVA_11
    export PATH=$JAVA_HOME/bin:$PATH

    cd module.tmp
    jar uf ${jarName} --release 9 -C v9 .
    if [ -d v11 ]
    then
        jar uf ${jarName} --release 11 -C v11 .
    fi
)
# Java 15 Step
(
    export JAVA_HOME=$JAVA_15
    export PATH=$JAVA_HOME/bin:$PATH

    cd module.tmp
    if [ -d v15 ]
    then
        jar uf ${jarName} --release 17 -C v15 .
    fi
)
cp module.tmp/${jarName} build/artifacts/jdk1.8/jars/$jarName

#bcsign11 build/artifacts/jdk1.8/jars/$jarName
done
```

It appears that at least `module-info.class` gets added now:
```
$ jar tf ~/src/bc-java/build/artifacts/jdk1.8/jars/bcprov-ext-jdk18on-171.jar| grep module
META-INF/versions/11/module-info.class
META-INF/versions/17/module-info.class
$ jar tf ~/src/bc-java/build/artifacts/jdk1.8/jars/bcprov-jdk18on-171.jar| grep module    
META-INF/versions/9/module-info.class
$ 
```

Here's the complete log, including the tests: 
[bc-log2.txt.gz](https://github.com/bcgit/bc-java/files/8084681/bc-log2.txt.gz)

After installation, I'm getting this:
```
$ java --list-modules -p /Library/Java/Extensions
.  .  .  .  .
org.bouncycastle.mail@171 file:///Library/Java/Extensions/bcmail-jdk18on-171.jar automatic
org.bouncycastle.pg@171 file:///Library/Java/Extensions/bcpg-jdk18on-171.jar automatic
org.bouncycastle.pkix@171 file:///Library/Java/Extensions/bcpkix-jdk18on-171.jar automatic
org.bouncycastle.provider file:///Library/Java/Extensions/bcprov-ext-jdk18on-171.jar
org.bouncycastle.tls@171 file:///Library/Java/Extensions/bctls-jdk18on-171.jar automatic
org.bouncycastle.util@171 file:///Library/Java/Extensions/bcutil-jdk18on-171.jar automatic
.  .  .  .  .
```
It means that only `bcprov-ext-...` got "modularized"? Do other jars need something similar???

I'd appreciate if you could show what should be changed/fixed in the above `add_module.sh` file, so that it's a "proper" module. And getting the tests to pass would be nice.

**Comment by dghgit** [comment]
Script kind of looks okay, suggest using -x (the shell trace mode) to see what it's actually doing.

**Comment by mouse07410** [comment]
> Script kind of looks okay, suggest using `-x` 

Understood, thank you!

Does the resulting module structure (see output below) look good to you?
```
$ jar tf ~/src/bc-java/build/artifacts/jdk1.8/jars/bcprov-ext-jdk18on-171.jar| grep module
META-INF/versions/11/module-info.class
META-INF/versions/17/module-info.class
$ jar tf ~/src/bc-java/build/artifacts/jdk1.8/jars/bcprov-jdk18on-171.jar| grep module    
META-INF/versions/9/module-info.class
$ 
$ java --list-modules -p /Library/Java/Extensions
.  .  .  .  .
org.bouncycastle.mail@171 file:///Library/Java/Extensions/bcmail-jdk18on-171.jar automatic
org.bouncycastle.pg@171 file:///Library/Java/Extensions/bcpg-jdk18on-171.jar automatic
org.bouncycastle.pkix@171 file:///Library/Java/Extensions/bcpkix-jdk18on-171.jar automatic
org.bouncycastle.provider file:///Library/Java/Extensions/bcprov-ext-jdk18on-171.jar
org.bouncycastle.tls@171 file:///Library/Java/Extensions/bctls-jdk18on-171.jar automatic
org.bouncycastle.util@171 file:///Library/Java/Extensions/bcutil-jdk18on-171.jar automatic
.  .  .  .  .
```

**Comment by dghgit** [comment]
I think the test failures are now sorted out. It took a bit...

**Comment by mouse07410** [comment]
Yes, I confirm. Thanks!
